cpp/src/arrow/compute/kernels/isin.cc

```diff
@@ -0,0 +1,325 @@
+  int64_t left_null_count{};
+  int64_t right_null_count{};
+  BooleanBuilder bool_builder_;
+  std::shared_ptr<ArrayData> output;
+};
+
+// ----------------------------------------------------------------------
+// Kernel wrapper for generic hash table kernels
+
+template <typename Type, typename Enable = void>
+struct IsInKernelTraits {};
+
+template <typename Type>
+struct IsInKernelTraits<Type, enable_if_null<Type>> {
+  using IsInKernelImpl = NullIsInKernel;
+};
+
+template <typename Type>
+struct IsInKernelTraits<Type, enable_if_has_c_type<Type>> {
+  using IsInKernelImpl = IsInKernel<Type, typename Type::c_type>;
+};
+
+template <typename Type>
+struct IsInKernelTraits<Type, enable_if_boolean<Type>> {
+  using IsInKernelImpl = IsInKernel<Type, bool>;
+};
+
+template <typename Type>
+struct IsInKernelTraits<Type, enable_if_binary<Type>> {
+  using IsInKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+template <typename Type>
+struct IsInKernelTraits<Type, enable_if_fixed_size_binary<Type>> {
+  using IsInKernelImpl = IsInKernel<Type, util::string_view>;
+};
+
+Status GetIsInKernel(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+                     std::unique_ptr<IsInKernelImpl>* out) {
+  std::unique_ptr<IsInKernelImpl> kernel;
+
+#define ISIN_CASE(InType)                                               \
+  case InType::type_id:                                                 \
+    kernel.reset(new typename IsInKernelTraits<InType>::IsInKernelImpl( \
+        type, ctx->memory_pool()));                                     \
+    break
+
+  switch (type->id()) {
+    ISIN_CASE(NullType);
+    ISIN_CASE(BooleanType);
+    ISIN_CASE(UInt8Type);
+    ISIN_CASE(Int8Type);
+    ISIN_CASE(UInt16Type);
+    ISIN_CASE(Int16Type);
+    ISIN_CASE(UInt32Type);
+    ISIN_CASE(Int32Type);
+    ISIN_CASE(UInt64Type);
+    ISIN_CASE(Int64Type);
+    ISIN_CASE(FloatType);
+    ISIN_CASE(DoubleType);
+    ISIN_CASE(Date32Type);
+    ISIN_CASE(Date64Type);
+    ISIN_CASE(Time32Type);
+    ISIN_CASE(Time64Type);
+    ISIN_CASE(TimestampType);
+    ISIN_CASE(BinaryType);
+    ISIN_CASE(StringType);
+    ISIN_CASE(FixedSizeBinaryType);
+    ISIN_CASE(Decimal128Type);
+    default:
+      break;
+  }
+#undef ISIN_CASE
+
+  if (!kernel) {
+    return Status::NotImplemented("isin", " not implemented for ", type->ToString());
+  }
+  RETURN_NOT_OK(kernel->Reset());
+  *out = std::move(kernel);
+  return Status::OK();
+}
+
+Status IsIn(FunctionContext* ctx, const Datum& left, const Datum& right, Datum* out) {
+  DCHECK(left.type()->Equals(right.type()));
+  std::vector<Datum> outputs;
+  std::unique_ptr<IsInKernelImpl> lkernel;
+
+  RETURN_NOT_OK(GetIsInKernel(ctx, left.type(), &lkernel));
+  RETURN_NOT_OK(lkernel->ConstructRight(ctx, right));
+
+  detail::PrimitiveAllocatingUnaryKernel kernel(lkernel.get());
+
+  RETURN_NOT_OK(detail::InvokeUnaryArrayKernel(ctx, &kernel, left, &outputs));
+
+  *out = detail::WrapDatumsLike(left, outputs);
```

Comment:
We can do it in a separate PR but it seems like we should have a version of InvokeUnaryArrayKernel that does this.